core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -293,14 +289,18 @@ public void runInContext() {
     // they receive cancel before start. Issue #1343 has more details
 
     // Propagate later Context cancellation to the remote side.
-    context.addListener(cancellationListener, directExecutor());
     if (effectiveDeadline != null
         // If the context has the effective deadline, we don't need to schedule an extra task.
         && !effectiveDeadline.equals(context.getDeadline())
         // If the channel has been terminated, we don't need to schedule an extra task.
-        && deadlineCancellationExecutor != null) {
-      deadlineCancellationFuture = startDeadlineTimer(effectiveDeadline);
+        && deadlineCancellationExecutor != null
+        // if already expired deadline let failing stream handle
+        && !(stream instanceof FailingClientStream)) {
+      deadlineCancellationNotifyApplicationFuture =
+          startDeadlineNotifyApplicationTimer(effectiveDeadline, observer);
     }
+    cancellationListener = new ContextCancellationListener(observer);
```

Comment:
I don't think this is thread-safe, since this field is read in `removeContextListenerAndCancelDeadlineFuture()`. I think you need to just move this initialization up higher before stream.start().